src/utils/utils_concurrency.h

```diff
@@ -92,57 +101,162 @@ static __inline unsigned char utils_mssb_index(long long value) {
-    InterlockedExchange64((LONG64 volatile *)object, (LONG64)desired)
+static __inline void utils_atomic_load_acquire_ptr(void **ptr, void **out) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    uintptr_t ret = (uintptr_t)InterlockedCompareExchangePointer(ptr, 0, 0);
+    utils_annotate_acquire((void *)ptr);
+    *(uintptr_t *)out = ret;
+}
 
-#define utils_atomic_increment(object)                                         \
-    InterlockedIncrement64((LONG64 volatile *)object)
+static __inline void utils_atomic_store_release_u64(uint64_t *ptr,
+                                                    uint64_t *val) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    ASSERT_IS_ALIGNED((uintptr_t)val, 8);
+    utils_annotate_release(ptr);
+    InterlockedExchange64((LONG64 volatile *)ptr, *(LONG64 *)val);
+}
 
-#define utils_atomic_decrement(object)                                         \
-    InterlockedDecrement64((LONG64 volatile *)object)
+static __inline void utils_atomic_store_release_ptr(void **ptr, void *val) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    utils_annotate_release(ptr);
+    InterlockedExchangePointer(ptr, val);
+}
 
-#define utils_fetch_and_add64(ptr, value)                                      \
-    InterlockedExchangeAdd64((LONG64 *)(ptr), value)
+static __inline uint64_t utils_atomic_increment_u64(uint64_t *ptr) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    // return incremented value
+    return InterlockedIncrement64((LONG64 volatile *)ptr);
+}
+
+static __inline uint64_t utils_atomic_decrement_u64(uint64_t *ptr) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    // return decremented value
+    return InterlockedDecrement64((LONG64 volatile *)ptr);
+}
+
+static __inline uint64_t utils_fetch_and_add_u64(uint64_t *ptr, uint64_t val) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    ASSERT_IS_ALIGNED((uintptr_t)&val, 8);
+    // return the value that had previously been in *ptr
+    return InterlockedExchangeAdd64((LONG64 volatile *)(ptr), val);
+}
+
+static __inline uint64_t utils_fetch_and_sub_u64(uint64_t *ptr, uint64_t val) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    ASSERT_IS_ALIGNED((uintptr_t)&val, 8);
+    // return the value that had previously been in *ptr
+    // NOTE: on Windows there is no *Sub* version of InterlockedExchange
+    return InterlockedExchangeAdd64((LONG64 volatile *)(ptr), -(LONG64)val);
+}
 
-// NOTE: windows version have different order of args
-#define utils_compare_exchange(object, desired, expected)                      \
-    InterlockedCompareExchange64((LONG64 volatile *)object, *expected, *desired)
+static __inline bool utils_compare_exchange_u64(uint64_t *ptr,
+                                                uint64_t *expected,
+                                                uint64_t *desired) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    ASSERT_IS_ALIGNED((uintptr_t)desired, 8);
+    ASSERT_IS_ALIGNED((uintptr_t)expected, 8);
```

Comment:
Are you sure that this ptrs has to by 8b aligned?

Reply:
https://learn.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedcompareexchange64

The variables for this function must be aligned on a 64-bit boundary; otherwise, this function will behave unpredictably on multiprocessor x86 systems and any non-x86 systems

Reply:
We are dereferencing this pointers, so it do not matter what alignment, it has here, as anyway it their value will be copied to the register/stack before call.

Reply:
Are you 100% sure that the compiler will copy these arguments and not use them directly? By the way, I really don't understand why we're discussing asserts here. In my opinion, there can never be too many asserts in the code.

Reply:
Guys, this function, takes second and third argument as a *Value*. So this do not matter PTR is aligned or not it do not matter at all, as function will get a *copy* values instead of the address. 

``` c
LONG64 InterlockedCompareExchange64(
  [in, out] LONG64 volatile *Destination,
  [in]      LONG64          ExChange,
  [in]      LONG64          Comperand
);```

Reply:
ahh, sorry, I did not saw that your comment was not about destination. Of coarse, only destination should be aligned.